[5.2] Adding a helper functionality to get the array key for a given value

ramirovjr opened the pull request:
- Added the method key() to the Illuminate\Support\Arr class
- Added a helper array_key to the Illuminate\Support\helpers function file

Imagine for instance that I have class with a constant using an array or a static class variable as an array and I want to get the array key using the value, as in the example below:

``` php
<?php

class Example {
    const TYPES = ['Free', 'Paid', 'Student'];
}

echo array_key(Example::TYPES, 'Paid'); // output: 1
echo Illuminate\Support\Arr::key(Example::TYPES, 'Student'); // output: 2
```

I have added the tests for the class and helper and documented both. Need to add now it to the framework documentation, which I believe is in a different repo, right?

I added this funcionality because I think it might be useful to others as it was for me today. Although it's up to you to decide it.

Ping me for any questions or need of modification to this PR.

Thanks!

ramirovarandas committed:
Adding a helper functionality to get the array key for a given value

- Added the method key() to the Illuminate\Support\Arr class
- Added a helper array_key to the Illuminate\Support\helpers function file

ramirovarandas committed:
Merge remote-tracking branch 'upstream/5.2' into 5.2

lucasmichot commented on src/Illuminate/Support/Arr.php:
```diff
@@ -525,4 +525,28 @@ public static function where($array, callable $callback)
+     */
+    public static function key($array, $value)
+    {
+        $lookFor = $value;
```
Why don't you just use http://php.net/manual/en/function.array-search.php ?

lucasmichot replied:
This pass your tests

```php
if ($keys = array_search($value, $array, true)) {
    return $keys;
}

return;
```

thomasruiz replied:
@lucasmichot is right. using `array_search` makes more sense. I'd suggest to return `false` explicitely rather than `null`.

lucasmichot commented on src/Illuminate/Support/Arr.php:
```diff
@@ -525,4 +525,28 @@ public static function where($array, callable $callback)
+     * @param mixed $value
+     * @return mixed
+     */
+    public static function key($array, $value)
```
`search` would make more sense

ramirovjr commented:
Ok, I'll change to array_search and also return false when not found.

ramirovarandas committed:
Adjusting the code for Arr::key to use array_search()

- Previously the code was using Arr::where internally and as suggested
through the comments, it was changed to PHP array_search()
- Keept the name Arr::key() and array_key(), because changing to Arr::search()
would imply changing array_key() to array_search(), which would colide with
current PHP internal function
- Adjusted the tests

ramirovjr commented:
As explained in the commit message, I changed internally to use array_search().
Keept the name Arr::key() and array_key() to avoid collision with PHP internal function array_search().

If accepted, I'll add it to the documentation. It's a separated repo, right?

lucasmichot commented:
Yep, https://github.com/laravel/docs

ramirovjr commented:
Thanks, I'll fork it and add the documentation if the PR is accepted.

lucasmichot commented on tests/Support/SupportArrTest.php:
```diff
@@ -461,4 +461,26 @@ public function testForget()
+        // Does not work for nested arrays
+        $array = ['emails' => ['joe@example.com' => ['name' => 'Joe'], 'jane@localhost' => ['name' => 'Jane']]];
+        $key = Arr::key($array, 'Jane');
+        $this->assertFalse($key);
```
What happens:

- if `$value` is an array of values
- If many keys are found
- If only one key is found
- If a same value is set for many keys

ramirovjr commented:
@lucasmichot This is my thought about this functionality:

- Return single value
- If multiple keys are found for the same value, return the first one

My idea when adding this item was based on some Ruby (and Rails) that I'm current having, where there you have:

```ruby
class Example
  TYPES = ['Free', 'Paid', 'Student']
end

Example::TYPES.index('Paid')

# Rails gives me the safe_at method

Example::TYPES.safe_at(2) # will return 'Student'
```

But of course, we must think towards to the Laravel and PHP community about the functionality. If it needs a change, no problem.

I agree that I need to add a test to check if it return the first key for multiple values. I'll add it.

ramirovarandas committed:
Adding test to verify that multiple identical values always return the first key

- When searching an array with multiple identical values, the first key must be
returned
- Added the test to both test files

ramirovarandas committed:
Merge remote-tracking branch 'upstream/5.2' into 5.2

vlakoff commented:
Now, it's strictly the same as simply using `array_search()`.

ramirovjr commented:
Strictly saying I agree, but it also make easier for new starters using the framework and also apply consistency for function/method usage, as we know that PHP (sometimes) makes us crazy with needle/haystack

taylorotwell commented:
Holding off on this.
